### PR TITLE
Add support for mean average precision (MAP)

### PR DESCRIPTION
**Pull request by Tigxy**

### CODE REVIEW

File: src/rmet/metrics.py [hunk 1]
```diff
@@ -150,6 +151,38 @@ def hitrate(logits: torch.Tensor, targets: torch.Tensor, k: int = 10, logits_are
     return recall
 
 
+def mean_average_precision(logits: torch.Tensor, targets: torch.Tensor, k: int = 10,
```
Comment (mmosc) [inline src/rmet/metrics.py:154]: If I understand correctly, this is still not averaged over users, so the returned tensor will be of shape `(n_samples, )`. So maybe it would be good to call it `average_precision` only?

File: src/rmet/metrics.py [hunk 2]
```diff
@@ -150,6 +151,38 @@ def hitrate(logits: torch.Tensor, targets: torch.Tensor, k: int = 10, logits_are
...
+                           logits_are_top_indices: bool = False):
+    """
+    Computes the mean_average_precision@k (MAP@k) for items.
+    In short, it combines precision values at all possible recall levels.
+
+    :param logits: prediction matrix about item relevance
+    :param targets: 0/1 matrix encoding true item relevance, same shape as logits
+    :param k: top k items to consider
+    :param logits_are_top_indices: whether logits are already top-k sorted indices
```
Comment (mmosc) [inline src/rmet/metrics.py:163]: Maybe add `:returns avg_precision:`?

File: src/rmet/metrics.py [hunk 3]
```diff
@@ -150,6 +151,38 @@ def hitrate(logits: torch.Tensor, targets: torch.Tensor, k: int = 10, logits_are
...
+    """
+    if k <= 0:
+        raise ValueError("k is required to be positive!")
+
+    top_indices = _get_top_k(logits, k, logits_are_top_indices, sorted=False)  # (n_samples, k)
```
Comment (mmosc) [inline src/rmet/metrics.py:168]: The rest of the code assumes that `top_indices` is sorted in descending order of logits, because it iterates over different cutoffs`ki`. If it is like that, then everything looks fine :)

**Review by mmosc — COMMENTED** [review]
Hey,

Thank you for adding mAP to this very nice repo! I left you some comments on its implementation, I hope they help :)

File: src/rmet/metrics.py (hunk 3 above)
Comment (Tigxy) [reply src/rmet/metrics.py:168]: Great catch! I'll update this to have the indices sorted :)

File: src/rmet/metrics.py (hunk 1 above)
Comment (Tigxy) [reply src/rmet/metrics.py:154]: Good point. I was thinking of calling it `mean_average_precision`, because the aggregated, i.e., averaged results would be `MAP` and not `AP`. For correctness, I'll change the name `average_precision` and will note this in the `README.md`.
